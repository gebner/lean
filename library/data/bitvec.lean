/-
Copyright (c) 2015 Joe Hendrix. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Joe Hendrix

Basic operations on bitvectors.

This is a work-in-progress, and contains additions to other theories.
-/
import data.tuple

def bitvec (n : ℕ) := tuple bool n

namespace bitvec
open nat
open tuple

instance (n : nat) : decidable_eq (bitvec n) :=
begin unfold bitvec, apply_instance end

-- Create a zero bitvector
def zero {n : ℕ} : bitvec n := tuple.repeat ff

-- Create a bitvector with the constant one.
def one {n : ℕ} : bitvec (succ n) := tuple.append (tuple.repeat ff : bitvec n) [ tt ]

section shift

  -- shift left
  def shl {n : ℕ} : bitvec n → ℕ → bitvec n
    | x i :=
       if le : i ≤ n then
         let eq := nat.sub_add_cancel le in
         cast (congr_arg bitvec eq) (tuple.append (dropn i x) (zero : bitvec i))
       else
         zero

  -- unsigned shift right
  def ushr {n : ℕ} : bitvec n → ℕ → bitvec n
    | x i :=
      if le : i ≤ n then
        let y : bitvec (n-i) := @firstn _ _ _ (sub_le n i) x in
        let eq := calc (i+(n-i)) = (n - i) + i : nat.add_comm i (n - i)
                            ...  = n           : nat.sub_add_cancel le in
        cast (congr_arg bitvec eq) (tuple.append zero y)
      else
        zero

  -- signed shift right
  def sshr {m : ℕ} : bitvec (succ m) → ℕ → bitvec (succ m)
    | x i :=
      let n := succ m in
      if le : i ≤ n then
        let z : bitvec i := repeat (head x) in
        let y : bitvec (n-i) := @firstn _ _ (n - i) (sub_le n i) x in
        let eq := calc (i+(n-i)) = (n-i) + i : nat.add_comm i (n - i)
                            ...  = n         : nat.sub_add_cancel le in
        cast (congr_arg bitvec eq) (tuple.append z y)
      else
        zero

end shift

section bitwise
  variable {n : ℕ}

  def not : bitvec n → bitvec n := map bnot
  def and : bitvec n → bitvec n → bitvec n := map₂ band
  def or  : bitvec n → bitvec n → bitvec n := map₂ bor
  def xor : bitvec n → bitvec n → bitvec n := map₂ bxor

end bitwise

section arith
  variable {n : ℕ}

  protected def xor3 (x:bool) (y:bool) (c:bool) := bxor (bxor x y) c
  protected def carry (x:bool) (y:bool) (c:bool) :=
    x && y || x && c || y && c

  def neg : bitvec n → bitvec n
  | x :=
    let f := λy c, (y || c, bxor y c) in
    prod.snd (map_accumr f x ff)

  -- Add with carry (no overflow)
  def adc : bitvec n → bitvec n → bool → bitvec (n+1)
  | x y c :=
    let f := λx y c, (bitvec.carry x y c, bitvec.xor3 x y c) in
    let z := tuple.map_accumr₂ f x y c in
    prod.fst z :: prod.snd z

  def add : bitvec n → bitvec n → bitvec n
  | x y := tail (adc x y ff)

  protected def borrow (x:bool) (y:bool) (b:bool) :=
    bnot x && y || bnot x && b || y && b

  -- Subtract with borrow
  def sbb : bitvec n → bitvec n → bool → bool × bitvec n
  | x y b :=
    let f := λx y c, (bitvec.borrow x y c, bitvec.xor3 x y c) in
    tuple.map_accumr₂ f x y b

  def sub : bitvec n → bitvec n → bitvec n
  | x y := prod.snd (sbb x y ff)

  instance : has_zero (bitvec n) := ⟨zero⟩
  /- Remark: we used to have the instance has_one only for (bitvec (succ n)).
     Now, we define it for all n to make sure we don't have to solve unification problems such as:
     succ ?n =?= (bit0 (bit1 one)) -/
  instance : has_one (bitvec n) :=
  match n with
  | 0            := ⟨zero⟩
  | (nat.succ n) := ⟨one⟩
  end
  instance : has_add (bitvec n)  := ⟨add⟩
  instance : has_sub (bitvec n)  := ⟨sub⟩
  instance : has_neg (bitvec n)  := ⟨neg⟩

  def mul : bitvec n → bitvec n → bitvec n
  | x y :=
    let f := λr b, cond b (r + r + y) (r + r) in
    list.foldl f 0 (to_list x)

  instance : has_mul (bitvec n)  := ⟨mul⟩
end arith

section comparison
  variable {n : ℕ}

  def uborrow : bitvec n → bitvec n → bool := λx y, prod.fst (sbb x y ff)

  def ult (x y : bitvec n) : Prop := uborrow x y = tt
  def ugt (x y : bitvec n) : Prop := ult y x

  def ule (x y : bitvec n) : Prop := ¬ (ult y x)
  def uge : bitvec n → bitvec n → Prop := λx y, ule y x

  def sborrow : bitvec (succ n) → bitvec (succ n) → bool := λx y,
     bool.cases_on
       (head x)
       (bool.cases_on (head y) (uborrow (tail x) (tail y)) tt)
       (bool.cases_on (head y) ff (uborrow (tail x) (tail y)))

  def slt : bitvec (succ n) → bitvec (succ n) → Prop := λx y,
     sborrow x y = tt

  def sgt : bitvec (succ n) → bitvec (succ n) → Prop := λx y, slt y x
  def sle : bitvec (succ n) → bitvec (succ n) → Prop := λx y, ¬ (slt y x)
  def sge : bitvec (succ n) → bitvec (succ n) → Prop := λx y, sle y x

end comparison

section from_bitvec
  variable {α : Type}

  -- Convert a bitvector to another number.
  def from_bitvec [has_add α] [has_zero α] [has_one α] {n : nat} (v : bitvec n) : α :=
  let f : α → bool → α := λr b, cond b (r + r + 1) (r + r) in
  list.foldl f (0 : α) (to_list v)
end from_bitvec

private def to_string {n : nat} : bitvec n → string
| ⟨bs, p⟩ :=
  "0b" ++ (bs^.reverse^.for (λ b, if b then #"1" else #"0"))

instance (n : nat) : has_to_string (bitvec n) :=
⟨to_string⟩
end bitvec
