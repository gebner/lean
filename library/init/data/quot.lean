/-
Copyright (c) 2015 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura

Quotient types.
-/
prelude
/- We import propext here, otherwise we would need a quot.lift for propositions. -/
import init.data.sigma.basic init.logic init.propext init.data.setoid

universes u v

-- iff can now be used to do substitutions in a calculation
attribute [subst]
lemma iff_subst {a b : Prop} {p : Prop → Prop} (h₁ : a ↔ b) (h₂ : p a) : p b :=
eq.subst (propext h₁) h₂

constant quotient {α : Sort u} (s : setoid α) : Sort u
protected constant quotient.mk {α : Sort u} [s : setoid α] : α → quotient s
protected axiom quotient.sound : ∀ {α : Sort u} [s : setoid α] {a b : α}, a ≈ b → quotient.mk a = quotient.mk b
protected constant quotient.rec {α : Sort u} [s : setoid α] {β : quotient s → Sort v}
    (f : ∀ a, β (quotient.mk a)) (h : ∀ a b, ∀ hs: a ≈ b, f a == f b) : ∀ q, β q

computation_rule quotient.comp_rule {α : Sort u} (a : α) [s : setoid α] {β : quotient s → Sort v}
    (f : ∀ a, β (quotient.mk a)) (h) : quotient.rec f h (quotient.mk a) = f a

attribute [elab_as_eliminator] quotient.rec

private def quotient.impl (α : Sort u) [setoid α] : Sort u := α
private def quotient.impl.mk {α : Sort u} [setoid α] : α → quotient.impl α := id
private def quotient.impl.rec {α : Sort u} [setoid α] {β : quotient.impl α → Sort v}
    (f : ∀ a, β (quotient.impl.mk a))
    (h : ∀ a b, ∀ hs : a ≈ b, f a == f b) : ∀ q, β q := f

unsafe_make_computable quotient.mk quotient.impl.mk
unsafe_make_computable quotient.rec quotient.impl.rec

namespace quotient

notation `⟦`:max a `⟧`:0 := quotient.mk a

attribute [reducible, elab_as_eliminator]
protected def lift {α : Sort u} {β : Sort v} [s : setoid α] (f : α → β) : (∀ a b, a ≈ b → f a = f b) → quotient s → β
| h q := quotient.rec f (λ a b hs, heq_of_eq (h a b hs)) q

attribute [elab_as_eliminator]
protected lemma ind {α : Sort u} [s : setoid α] {β : quotient s → Prop} : (∀ a, β ⟦a⟧) → ∀ q, β q
| h q := quotient.rec h (λ a b hs, heq_of_eq_rec_left (quotient.sound hs) rfl) q

attribute [reducible, elab_as_eliminator]
protected def lift_on {α : Sort u} {β : Sort v} [s : setoid α] (q : quotient s) (f : α → β) (c : ∀ a b, a ≈ b → f a = f b) : β :=
quotient.lift f c q

attribute [elab_as_eliminator]
protected lemma induction_on {α : Sort u} [s : setoid α] {β : quotient s → Prop} (q : quotient s) (h : ∀ a, β ⟦a⟧) : β q :=
quotient.ind h q

lemma exists_rep {α : Sort u} [s : setoid α] (q : quotient s) : ∃ a : α, ⟦a⟧ = q :=
quotient.induction_on q (λ a, ⟨a, rfl⟩)

section
variable {α : Sort u}
variable [s : setoid α]
variable {β : quotient s → Sort v}

attribute [reducible, elab_as_eliminator]
protected def rec_on
   (q : quotient s) (f : Π a, β ⟦a⟧) (h : ∀ (a b : α) (p : a ≈ b), f a == f b) : β q :=
quotient.rec f h q

attribute [reducible, elab_as_eliminator]
protected def rec_on_subsingleton
   [h : ∀ a, subsingleton (β ⟦a⟧)] (q : quotient s) (f : Π a, β ⟦a⟧) : β q :=
quotient.rec_on q f (λ a b hs, subsingleton.helim (quotient.sound hs ▸ rfl) _ _)

end

section
universes u_a u_b u_c
variables {α : Sort u_a} {β : Sort u_b} {φ : Sort u_c}
variables [s₁ : setoid α] [s₂ : setoid β]
include s₁ s₂

attribute [reducible, elab_as_eliminator]
protected def lift₂
   (f : α → β → φ)(c : ∀ a₁ a₂ b₁ b₂, a₁ ≈ b₁ → a₂ ≈ b₂ → f a₁ a₂ = f b₁ b₂)
   (q₁ : quotient s₁) (q₂ : quotient s₂) : φ :=
quotient.lift
  (λ (a₁ : α), quotient.lift (f a₁) (λ (a b : β), c a₁ a a₁ b (setoid.refl a₁)) q₂)
  (λ (a b : α) (h : a ≈ b),
     @quotient.ind β s₂
       (λ (a_1 : quotient _),
          (quotient.lift (f a) (λ (a_1 b : β), c a a_1 a b (setoid.refl a)) a_1)
          =
          (quotient.lift (f b) (λ (a b_1 : β), c b a b b_1 (setoid.refl b)) a_1))
       (λ (a' : β), c a a' b a' h (setoid.refl a'))
       q₂)
  q₁

attribute [reducible, elab_as_eliminator]
protected def lift_on₂
  (q₁ : quotient s₁) (q₂ : quotient s₂) (f : α → β → φ) (c : ∀ a₁ a₂ b₁ b₂, a₁ ≈ b₁ → a₂ ≈ b₂ → f a₁ a₂ = f b₁ b₂) : φ :=
quotient.lift₂ f c q₁ q₂

attribute [elab_as_eliminator]
protected lemma ind₂ {φ : quotient s₁ → quotient s₂ → Prop} (h : ∀ a b, φ ⟦a⟧ ⟦b⟧) (q₁ : quotient s₁) (q₂ : quotient s₂) : φ q₁ q₂ :=
quotient.ind (λ a₁, quotient.ind (λ a₂, h a₁ a₂) q₂) q₁

attribute [elab_as_eliminator]
protected lemma induction_on₂
   {φ : quotient s₁ → quotient s₂ → Prop} (q₁ : quotient s₁) (q₂ : quotient s₂) (h : ∀ a b, φ ⟦a⟧ ⟦b⟧) : φ q₁ q₂ :=
quotient.ind (λ a₁, quotient.ind (λ a₂, h a₁ a₂) q₂) q₁

attribute [elab_as_eliminator]
protected lemma induction_on₃
   [s₃ : setoid φ]
   {δ : quotient s₁ → quotient s₂ → quotient s₃ → Prop} (q₁ : quotient s₁) (q₂ : quotient s₂) (q₃ : quotient s₃) (h : ∀ a b c, δ ⟦a⟧ ⟦b⟧ ⟦c⟧)
   : δ q₁ q₂ q₃ :=
quotient.ind (λ a₁, quotient.ind (λ a₂, quotient.ind (λ a₃, h a₁ a₂ a₃) q₃) q₂) q₁
end

section exact
variable   {α : Sort u}
variable   [s : setoid α]
include s

private def rel (q₁ q₂ : quotient s) : Prop :=
quotient.lift_on₂ q₁ q₂
  (λ a₁ a₂, a₁ ≈ a₂)
  (λ a₁ a₂ b₁ b₂ a₁b₁ a₂b₂,
    propext (iff.intro
      (λ a₁a₂, setoid.trans (setoid.symm a₁b₁) (setoid.trans a₁a₂ a₂b₂))
      (λ b₁b₂, setoid.trans a₁b₁ (setoid.trans b₁b₂ (setoid.symm a₂b₂)))))

local infix `~` := rel

private lemma rel.refl : ∀ q : quotient s, q ~ q :=
λ q, quotient.induction_on q (λ a, setoid.refl a)

private lemma eq_imp_rel {q₁ q₂ : quotient s} : q₁ = q₂ → q₁ ~ q₂ :=
assume h, eq.rec_on h (rel.refl q₁)

lemma exact {a b : α} : ⟦a⟧ = ⟦b⟧ → a ≈ b :=
assume h, eq_imp_rel h
end exact

section
universes u_a u_b u_c
variables {α : Sort u_a} {β : Sort u_b}
variables [s₁ : setoid α] [s₂ : setoid β]
include s₁ s₂

attribute [reducible, elab_as_eliminator]
protected def rec_on_subsingleton₂
   {φ : quotient s₁ → quotient s₂ → Sort u_c} [h : ∀ a b, subsingleton (φ ⟦a⟧ ⟦b⟧)]
   (q₁ : quotient s₁) (q₂ : quotient s₂) (f : Π a b, φ ⟦a⟧ ⟦b⟧) : φ q₁ q₂:=
@quotient.rec_on_subsingleton _ s₁ (λ q, φ q q₂) (λ a, quotient.ind (λ b, h a b) q₂) q₁
  (λ a, quotient.rec_on_subsingleton q₂ (λ b, f a b))

end
end quotient

open decidable
instance {α : Sort u} {s : setoid α} [d : ∀ a b : α, decidable (a ≈ b)] : decidable_eq (quotient s) :=
λ q₁ q₂,
  quotient.rec_on_subsingleton₂ q₁ q₂
    (λ a₁ a₂,
      match (d a₁ a₂) with
      | (is_true h₁)  := is_true (quotient.sound h₁)
      | (is_false h₂) := is_false (λ h, h₂ (quotient.exact h))
      end)
