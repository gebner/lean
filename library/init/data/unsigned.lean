/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import init.data.fin.basic

open nat
def unsigned_sz : nat := succ 4294967295

attribute [reducible]
def unsigned := fin unsigned_sz

namespace unsigned
/- We cannot use tactic dec_trivial here because the tactic framework has not been defined yet. -/
private lemma zero_lt_unsigned_sz : 0 < unsigned_sz :=
zero_lt_succ _

def of_nat (n : nat) : unsigned :=
if h : n < unsigned_sz then fin.mk n h else fin.mk 0 zero_lt_unsigned_sz

def to_nat (c : unsigned) : nat :=
fin.val c

def succ (i : unsigned) :=
of_nat i^.to_nat^.succ

end unsigned

instance : has_zero unsigned := ⟨unsigned.of_nat 0⟩

instance : decidable_eq unsigned :=
have decidable_eq (fin unsigned_sz), from fin.decidable_eq _,
this
