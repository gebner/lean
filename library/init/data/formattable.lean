/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
prelude
import init.data.string.basic init.data.bool.basic init.data.subtype.basic
import init.data.unsigned.basic init.data.prod init.data.sum.basic init.data.nat.div
import init.coe

universes u v

namespace nat

def digit_char (n : ℕ) : char :=
if n = 0 then '0' else
if n = 1 then '1' else
if n = 2 then '2' else
if n = 3 then '3' else
if n = 4 then '4' else
if n = 5 then '5' else
if n = 6 then '6' else
if n = 7 then '7' else
if n = 8 then '8' else
if n = 9 then '9' else
if n = 0xa then 'a' else
if n = 0xb then 'b' else
if n = 0xc then 'c' else
if n = 0xd then 'd' else
if n = 0xe then 'e' else
if n = 0xf then 'f' else
'*'

def digit_succ (base : ℕ) : list ℕ → list ℕ
| [] := [1]
| (d::ds) :=
    if d+1 = base then
        0 :: digit_succ ds
    else
        (d+1) :: ds

def to_digits (base : ℕ) : ℕ → list ℕ
| 0 := [0]
| (n+1) := digit_succ base (to_digits n)

protected def repr (n : ℕ) : string :=
((to_digits 10 n).map digit_char).reverse.as_string

end nat

class formattable (α : Type u) extends has_append α :=
(of_string' : string → α)
(line' : α := of_string' "\n")
(of_nat' : ℕ → α := λ i, of_string' i.repr)
(nest : ℕ → α → α := λ _, id)
(group : α → α := id)

instance : formattable string :=
{ of_string' := id, append := (++) }

namespace formattable
variables {α : Type u} [formattable α]

def of_nat (n : ℕ) : α :=
of_nat' _ n

instance lift_from_nat : has_lift ℕ α :=
⟨of_nat⟩

def of_string (s : string) : α :=
of_string' _ s

instance lift_from_string : has_lift string α :=
⟨of_string⟩

instance : inhabited α :=
⟨↑""⟩

def line : α :=
line' _

def space : α :=
↑" "

def indent (a : α) (i : ℕ) : α :=
nest i (line ++ a)

def join (xs : list α) : α :=
xs.foldl (++) ↑""

def bracket : string → string → α → α
| o c f := ↑o ++ nest (utf8_length o) f ++ ↑c

def paren (f : α) : α :=
bracket "(" ")" f

def cbrace (f : α) : α :=
bracket "{" "}" f

def sbracket (f : α) : α :=
bracket "[" "]" f

def dcbrace (f : α) : α :=
↑"⦃" ++ nest 1 f ++ ↑"⦄"

end formattable