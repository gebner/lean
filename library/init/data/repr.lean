/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
prelude
import init.data.string.basic init.data.bool.basic init.data.subtype.basic
import init.data.unsigned.basic init.data.prod init.data.sum.basic init.data.nat.div
import init.data.formattable
open sum subtype nat

universes u v

class has_repr (α : Type u) :=
(repr : α → string)

def repr {α : Type u} [has_repr α] : α → string :=
has_repr.repr

instance : has_repr bool :=
⟨λ b, cond b "tt" "ff"⟩

instance {p : Prop} : has_repr (decidable p) :=
-- Remark: type class inference will not consider local instance `b` in the new elaborator
⟨λ b : decidable p, @ite p b _ "tt" "ff"⟩

protected def list.repr_aux {α : Type u} [has_repr α] : bool → list α → string
| b  []      := ""
| tt (x::xs) := repr x ++ list.repr_aux ff xs
| ff (x::xs) := ", " ++ repr x ++ list.repr_aux ff xs

protected def list.repr {α : Type u} [has_repr α] : list α → string
| []      := "[]"
| (x::xs) := "[" ++ list.repr_aux tt (x::xs) ++ "]"

instance {α : Type u} [has_repr α] : has_repr (list α) :=
⟨list.repr⟩

instance : has_repr unit :=
⟨λ u, "star"⟩

instance {α : Type u} [has_repr α] : has_repr (option α) :=
⟨λ o, match o with | none := "none" | (some a) := "(some " ++ repr a ++ ")" end⟩

instance {α : Type u} {β : Type v} [has_repr α] [has_repr β] : has_repr (α ⊕ β) :=
⟨λ s, match s with | (inl a) := "(inl " ++ repr a ++ ")" | (inr b) := "(inr " ++ repr b ++ ")" end⟩

instance {α : Type u} {β : Type v} [has_repr α] [has_repr β] : has_repr (α × β) :=
⟨λ ⟨a, b⟩, "(" ++ repr a ++ ", " ++ repr b ++ ")"⟩

instance {α : Type u} {β : α → Type v} [has_repr α] [s : ∀ x, has_repr (β x)] : has_repr (sigma β) :=
⟨λ ⟨a, b⟩, "⟨"  ++ repr a ++ ", " ++ repr b ++ "⟩"⟩

instance {α : Type u} {p : α → Prop} [has_repr α] : has_repr (subtype p) :=
⟨λ s, repr (val s)⟩

instance : has_repr nat :=
⟨nat.repr⟩

def hex_digit_repr (n : nat) : string :=
string.singleton $ nat.digit_char n

def char_to_hex (c : char) : string :=
let n  := char.to_nat c,
    d2 := n / 16,
    d1 := n % 16
in hex_digit_repr d2 ++ hex_digit_repr d1

def char.quote_core (c : char) : string :=
if       c = '\n' then "\\n"
else if  c = '\t' then "\\t"
else if  c = '\\' then "\\\\"
else if  c = '\"' then "\\\""
else if  char.to_nat c <= 31 then "\\x" ++ char_to_hex c
else string.singleton c

instance : has_repr char :=
⟨λ c, "'" ++ char.quote_core c ++ "'"⟩

def string.quote_aux : list char → string
| []      := ""
| (x::xs) := char.quote_core x ++ string.quote_aux xs

def string.quote (s : string) : string :=
if s.is_empty = tt then "\"\""
else "\"" ++ string.quote_aux s.to_list ++ "\""

instance : has_repr string :=
⟨string.quote⟩

instance (n : nat) : has_repr (fin n) :=
⟨λ f, repr (fin.val f)⟩

instance : has_repr unsigned :=
⟨λ n, repr (fin.val n)⟩

def char.repr (c : char) : string :=
repr c
