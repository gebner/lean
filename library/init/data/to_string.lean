/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
prelude
import init.data.string.basic init.data.bool.basic init.data.subtype.basic
import init.data.unsigned.basic init.data.prod init.data.sum.basic init.data.nat.div
import init.data.repr
open sum subtype nat

universes u v w

section

class has_to_fmt (γ : Type w) (α : Type u) :=
(to_fmt : α → γ)

variables {γ : Type w} [formattable γ]

def to_fmt {α : Type u} [has_to_fmt γ α] : α → γ :=
has_to_fmt.to_fmt _

def to_string {α : Type u} [has_to_fmt string α] : α → string :=
to_fmt

instance has_to_fmt.id : has_to_fmt γ γ :=
⟨id⟩

instance : has_to_fmt γ string :=
⟨λ s, ↑s⟩

instance : has_to_fmt γ bool :=
⟨λ b, cond b ↑"tt" ↑"ff"⟩

instance {p : Prop} : has_to_fmt γ (decidable p) :=
-- Remark: type class inference will not consider local instance `b` in the new elaborator
⟨λ b : decidable p, @ite p b _ ↑"tt" ↑"ff"⟩

protected def list.to_string_aux {α : Type u} [has_to_fmt γ α] : bool → list α → γ
| b  []      := ↑""
| tt (x::xs) := to_fmt x ++ list.to_string_aux ff xs
| ff (x::xs) := ↑", " ++ to_fmt x ++ list.to_string_aux ff xs

protected def list.to_string {α : Type u} [has_to_fmt γ α] : list α → γ
| []      := ↑"[]"
| (x::xs) := ↑"[" ++ list.to_string_aux tt (x::xs) ++ ↑"]"

instance has_to_fmt_list {α : Type u} [has_to_fmt γ α] : has_to_fmt γ (list α) :=
⟨list.to_string⟩

instance : has_to_fmt γ unit :=
⟨λ u, ↑"()"⟩

instance : has_to_fmt γ nat :=
⟨λ n, formattable.of_nat n⟩

instance : has_to_fmt γ char :=
⟨λ c, ↑c.to_string⟩

instance (n : nat) : has_to_fmt γ (fin n) :=
⟨λ f, to_fmt (fin.val f)⟩

instance : has_to_fmt γ unsigned :=
⟨λ n, to_fmt (fin.val n)⟩

instance {α : Type u} [has_to_fmt γ α] : has_to_fmt γ (option α) :=
⟨λ o, match o with | none := to_fmt "none" | (some a) := to_fmt "(some " ++ to_fmt a ++ to_fmt ")" end⟩

instance {α : Type u} {β : Type v} [has_to_fmt γ α] [has_to_fmt γ β] : has_to_fmt γ (α ⊕ β) :=
⟨λ s, match s with | (inl a) := to_fmt "(inl " ++ to_fmt a ++ ↑")" | (inr b) := to_fmt "(inr " ++ to_fmt b ++ to_fmt ")" end⟩

instance {α : Type u} {β : Type v} [has_to_fmt γ α] [has_to_fmt γ β] : has_to_fmt γ (α × β) :=
⟨λ ⟨a, b⟩, to_fmt "(" ++ to_fmt a ++ ↑", " ++ to_fmt b ++ ↑")"⟩

instance {α : Type u} {β : α → Type v} [has_to_fmt γ α] [s : ∀ x, has_to_fmt γ (β x)] : has_to_fmt γ (sigma β) :=
⟨λ ⟨a, b⟩, ↑"⟨"  ++ to_fmt a ++ ↑", " ++ to_fmt b ++ ↑"⟩"⟩

instance {α : Type u} {p : α → Prop} [has_to_fmt γ α] : has_to_fmt γ (subtype p) :=
⟨λ s, to_fmt (val s)⟩

end