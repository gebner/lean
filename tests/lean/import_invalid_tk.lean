import data.bitvec 0b311

print bitvec
