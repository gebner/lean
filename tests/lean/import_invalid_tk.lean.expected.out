import_invalid_tk.lean:1:21: error: invalid binary digit
@[reducible]
def bitvec : ℕ → Type :=
λ (n : ℕ), tuple bool n
